---
title: "R Notebook for BCB546 R Assignment"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 



#Part I generating files for snps on each chromosome of maize and teosinte groups

#To read in the data files
```{r}
fang_et_al_genotypes <- read.delim("fang_et_al_genotypes.txt", header = TRUE)
fang_et_al_genotypes <- read.delim("fang_et_al_genotypes.txt", header = TRUE)
```

#To read in as a matrix
```{r}
snp_position <- read.delim ("snp_position.txt", header = TRUE, stringsAsFactors = FALSE)
```

#Converting the dataframes to a tibble, needed to use the dplyr package functions
```{r}
library(dplyr)
fang_df <- tbl_df(fang_et_al_genotypes)
snp_position_df <- tbl_df(snp_position)
```

#Converting the position column in snp_position to numeric
```{r}
snp_position_df <- transform(snp_position_df, Position = as.numeric(Position))
```

#Remove the desired maize or teosinte groups
```{r}
maize_target_groups <- c("ZMMIL", "ZMMLR","ZMMMR")
fang_maize <- filter(fang_df, Group %in% maize_target_groups)
teosinte_target_groups <- c("ZMPBA", "ZMPIL", "ZMPJA")
fang_teosinte <- filter(fang_df, Group %in% teosinte_target_groups)
```

#Transpose the group genotype files
```{r}
maize_trim <- fang_maize[, -c(1:3)]
fang_maize_t <- t(maize_trim)

teosinte_trim <- fang_teosinte[, -c(1:3)]
fang_teosinte_t <- t(teosinte_trim)
```

#To convert the row names to the first column being the SNP ID
```{r}
SNP_ID <- rownames(fang_maize_t)
rownames(fang_maize_t) <- NULL
fang_maize_ID <- cbind(SNP_ID,fang_maize_t)

SNP_ID <- rownames(fang_teosinte_t)
rownames(fang_teosinte_t) <- NULL
fang_teosinte_ID <- cbind(SNP_ID,fang_teosinte_t)
```

#Changing the tranformed matrix of snp_id to a data frame
```{r}
fang_maize_ID <- as.data.frame(fang_maize_ID, stringAsFactors = FALSE)
fang_teosinte_ID <- as.data.frame(fang_teosinte_ID, stringsAsFactor = FALSE)
```

#Join the genotype data file to the snp_position data file
```{r}
joined_maize <- left_join(snp_position_df, fang_maize_ID, by = c("SNP_ID"))
joined_teosinte <- left_join(snp_position_df, fang_teosinte_ID, by = c("SNP_ID"))
```

#Convert joined files to character
```{r}
joined_maize <- data.frame(lapply(joined_maize, as.character), stringsAsFactors=FALSE)
joined_teosinte <- data.frame(lapply(joined_teosinte, as.character), stringsAsFactors=FALSE)
```

#To remove each chromosome for data files with missing data encoded as ?/? 
```{r}
maize_chr1_question  <- filter(joined_maize, Chromosome == 1)
maize_chr2_question  <- filter(joined_maize, Chromosome == 2)
maize_chr3_question  <- filter(joined_maize, Chromosome == 3)
maize_chr4_question  <- filter(joined_maize, Chromosome == 4)
maize_chr5_question  <- filter(joined_maize, Chromosome == 5)
maize_chr6_question  <- filter(joined_maize, Chromosome == 6)
maize_chr7_question  <- filter(joined_maize, Chromosome == 7)
maize_chr8_question  <- filter(joined_maize, Chromosome == 8)
maize_chr9_question  <- filter(joined_maize, Chromosome == 9)
maize_chr10_question  <- filter(joined_maize, Chromosome == 10)

teosinte_chr1_question  <- filter(joined_teosinte, Chromosome == 1)
teosinte_chr2_question  <- filter(joined_teosinte, Chromosome == 2)
teosinte_chr3_question  <- filter(joined_teosinte, Chromosome == 3)
teosinte_chr4_question  <- filter(joined_teosinte, Chromosome == 4)
teosinte_chr5_question  <- filter(joined_teosinte, Chromosome == 5)
teosinte_chr6_question  <- filter(joined_teosinte, Chromosome == 6)
teosinte_chr7_question  <- filter(joined_teosinte, Chromosome == 7)
teosinte_chr8_question  <- filter(joined_teosinte, Chromosome == 8)
teosinte_chr9_question  <- filter(joined_teosinte, Chromosome == 9)
teosinte_chr10_question  <- filter(joined_teosinte, Chromosome == 10)
```



#To sort files based on increasing position and generate the final formate for the first set of files
```{r}
maize_chr1_incr <- arrange(maize_chr1_question,as.numeric(Position))
maize_chr2_incr <- arrange(maize_chr2_question,as.numeric(Position))
maize_chr3_incr <- arrange(maize_chr3_question,as.numeric(Position))
maize_chr4_incr <- arrange(maize_chr4_question,as.numeric(Position))
maize_chr5_incr <- arrange(maize_chr5_question,as.numeric(Position))
maize_chr6_incr <- arrange(maize_chr6_question,as.numeric(Position))
maize_chr7_incr <- arrange(maize_chr7_question,as.numeric(Position))
maize_chr8_incr <- arrange(maize_chr8_question,as.numeric(Position))
maize_chr9_incr <- arrange(maize_chr9_question,as.numeric(Position))
maize_chr10_incr <- arrange(maize_chr10_question,as.numeric(Position))

teosinte_chr1_incr <- arrange(teosinte_chr1_question,as.numeric(Position))
teosinte_chr2_incr <- arrange(teosinte_chr2_question,as.numeric(Position))
teosinte_chr3_incr <- arrange(teosinte_chr3_question,as.numeric(Position))
teosinte_chr4_incr <- arrange(teosinte_chr4_question,as.numeric(Position))
teosinte_chr5_incr <- arrange(teosinte_chr5_question,as.numeric(Position))
teosinte_chr6_incr <- arrange(teosinte_chr6_question,as.numeric(Position))
teosinte_chr7_incr <- arrange(teosinte_chr7_question,as.numeric(Position))
teosinte_chr8_incr <- arrange(teosinte_chr8_question,as.numeric(Position))
teosinte_chr9_incr <- arrange(teosinte_chr9_question,as.numeric(Position))
teosinte_chr10_incr <- arrange(teosinte_chr10_question,as.numeric(Position))
```

#To replace the question mark with a dash
```{r}
joined_maize[joined_maize=="?/?"]<- "-/-"
joined_teosinte[joined_teosinte=="?/?"]<- "-/-"
```

#To remove each chromosome for data files with missing data encoded as -/-
```{r}
maize_chr1_dash  <- filter(joined_maize, Chromosome == 1)
maize_chr2_dash  <- filter(joined_maize, Chromosome == 2)
maize_chr3_dash  <- filter(joined_maize, Chromosome == 3)
maize_chr4_dash  <- filter(joined_maize, Chromosome == 4)
maize_chr5_dash  <- filter(joined_maize, Chromosome == 5)
maize_chr6_dash  <- filter(joined_maize, Chromosome == 6)
maize_chr7_dash  <- filter(joined_maize, Chromosome == 7)
maize_chr8_dash  <- filter(joined_maize, Chromosome == 8)
maize_chr9_dash  <- filter(joined_maize, Chromosome == 9)
maize_chr10_dash  <- filter(joined_maize, Chromosome == 10)

teosinte_chr1_dash  <- filter(joined_teosinte, Chromosome == 1)
teosinte_chr2_dash  <- filter(joined_teosinte, Chromosome == 2)
teosinte_chr3_dash  <- filter(joined_teosinte, Chromosome == 3)
teosinte_chr4_dash  <- filter(joined_teosinte, Chromosome == 4)
teosinte_chr5_dash  <- filter(joined_teosinte, Chromosome == 5)
teosinte_chr6_dash  <- filter(joined_teosinte, Chromosome == 6)
teosinte_chr7_dash  <- filter(joined_teosinte, Chromosome == 7)
teosinte_chr8_dash  <- filter(joined_teosinte, Chromosome == 8)
teosinte_chr9_dash  <- filter(joined_teosinte, Chromosome == 9)
teosinte_chr10_dash  <- filter(joined_teosinte, Chromosome == 10)
```


#Sort dash files in decreasing order and generate the final formate for the second set of files
```{r}
maize_chr1_decr <- arrange(maize_chr1_dash,desc(as.numeric(Position)))
maize_chr2_decr <- arrange(maize_chr2_dash,desc(as.numeric(Position)))
maize_chr3_decr <- arrange(maize_chr3_dash,desc(as.numeric(Position)))
maize_chr4_decr <- arrange(maize_chr4_dash,desc(as.numeric(Position)))
maize_chr5_decr <- arrange(maize_chr5_dash,desc(as.numeric(Position)))
maize_chr6_decr <- arrange(maize_chr6_dash,desc(as.numeric(Position)))
maize_chr7_decr <- arrange(maize_chr7_dash,desc(as.numeric(Position)))
maize_chr8_decr <- arrange(maize_chr8_dash,desc(as.numeric(Position)))
maize_chr9_decr <- arrange(maize_chr9_dash,desc(as.numeric(Position)))
maize_chr10_decr <- arrange(maize_chr10_dash,desc(as.numeric(Position)))

teosinte_chr1_decr <- arrange(teosinte_chr1_dash,desc(as.numeric(Position)))
teosinte_chr2_decr <- arrange(teosinte_chr2_dash,desc(as.numeric(Position)))
teosinte_chr3_decr <- arrange(teosinte_chr3_dash,desc(as.numeric(Position)))
teosinte_chr4_decr <- arrange(teosinte_chr4_dash,desc(as.numeric(Position)))
teosinte_chr5_decr <- arrange(teosinte_chr5_dash,desc(as.numeric(Position)))
teosinte_chr6_decr <- arrange(teosinte_chr6_dash,desc(as.numeric(Position)))
teosinte_chr7_decr <- arrange(teosinte_chr7_dash,desc(as.numeric(Position)))
teosinte_chr8_decr <- arrange(teosinte_chr8_dash,desc(as.numeric(Position)))
teosinte_chr9_decr <- arrange(teosinte_chr9_dash,desc(as.numeric(Position)))
teosinte_chr10_decr <- arrange(teosinte_chr10_dash,desc(as.numeric(Position)))
```


#Write final files to .csv
```{r}
write.csv(maize_chr1_incr, "maize_chr1_increasing_order.csv", row.names=F)
write.csv(maize_chr2_incr, "maize_chr2_increasing_order.csv", row.names=F)
write.csv(maize_chr3_incr, "maize_chr3_increasing_order.csv", row.names=F)
write.csv(maize_chr4_incr, "maize_chr4_increasing_order.csv", row.names=F)
write.csv(maize_chr5_incr, "maize_chr5_increasing_order.csv", row.names=F)
write.csv(maize_chr6_incr, "maize_chr6_increasing_order.csv", row.names=F)
write.csv(maize_chr7_incr, "maize_chr7_increasing_order.csv", row.names=F)
write.csv(maize_chr8_incr, "maize_chr8_increasing_order.csv", row.names=F)
write.csv(maize_chr9_incr, "maize_chr9_increasing_order.csv", row.names=F)
write.csv(maize_chr10_incr, "maize_chr10_increasing_order.csv", row.names=F)

write.csv(teosinte_chr1_incr, "teosinte_chr1_increasing_order.csv", row.names=F)
write.csv(teosinte_chr2_incr, "teosinte_chr2_increasing_order.csv", row.names=F)
write.csv(teosinte_chr3_incr, "teosinte_chr3_increasing_order.csv", row.names=F)
write.csv(teosinte_chr4_incr, "teosinte_chr4_increasing_order.csv", row.names=F)
write.csv(teosinte_chr5_incr, "teosinte_chr5_increasing_order.csv", row.names=F)
write.csv(teosinte_chr6_incr, "teosinte_chr6_increasing_order.csv", row.names=F)
write.csv(teosinte_chr7_incr, "teosinte_chr7_increasing_order.csv", row.names=F)
write.csv(teosinte_chr8_incr, "teosinte_chr8_increasing_order.csv", row.names=F)
write.csv(teosinte_chr9_incr, "teosinte_chr9_increasing_order.csv", row.names=F)
write.csv(teosinte_chr10_incr, "teosinte_chr10_increasing_order.csv", row.names=F)

write.csv(maize_chr1_decr, "maize_chr1_decreasing_order.csv", row.names=F)
write.csv(maize_chr2_decr, "maize_chr2_decreasing_order.csv", row.names=F)
write.csv(maize_chr3_decr, "maize_chr3_decreasing_order.csv", row.names=F)
write.csv(maize_chr4_decr, "maize_chr4_decreasing_order.csv", row.names=F)
write.csv(maize_chr5_decr, "maize_chr5_decreasing_order.csv", row.names=F)
write.csv(maize_chr6_decr, "maize_chr6_decreasing_order.csv", row.names=F)
write.csv(maize_chr7_decr, "maize_chr7_decreasing_order.csv", row.names=F)
write.csv(maize_chr8_decr, "maize_chr8_decreasing_order.csv", row.names=F)
write.csv(maize_chr9_decr, "maize_chr9_decreasing_order.csv", row.names=F)
write.csv(maize_chr10_decr, "maize_chr10_decreasing_order.csv", row.names=F)

write.csv(teosinte_chr1_decr, "teosinte_chr1_decreasing_order.csv", row.names=F)
write.csv(teosinte_chr2_decr, "teosinte_chr2_decreasing_order.csv", row.names=F)
write.csv(teosinte_chr3_decr, "teosinte_chr3_decreasing_order.csv", row.names=F)
write.csv(teosinte_chr4_decr, "teosinte_chr4_decreasing_order.csv", row.names=F)
write.csv(teosinte_chr5_decr, "teosinte_chr5_decreasing_order.csv", row.names=F)
write.csv(teosinte_chr6_decr, "teosinte_chr6_decreasing_order.csv", row.names=F)
write.csv(teosinte_chr7_decr, "teosinte_chr7_decreasing_order.csv", row.names=F)
write.csv(teosinte_chr8_decr, "teosinte_chr8_decreasing_order.csv", row.names=F)
write.csv(teosinte_chr9_decr, "teosinte_chr9_decreasing_order.csv", row.names=F)
write.csv(teosinte_chr10_decr, "teosinte_chr10_decreasing_order.csv", row.names=F)
```


#PART II########################################################################
#Generating tidy data#########################################################
#Melt genotype data
```{r}
library(reshape2)
fang_trim <- fang_et_al_genotypes[, -c(1:3)]
fang_t <- t(fang_trim)
SNP_ID <- rownames(fang_t)
rownames(fang_t) <- NULL
fang_SNP_ID <- cbind(SNP_ID,fang_t)
fang_SNP_ID <- as.data.frame(fang_SNP_ID, stringAsFactors = FALSE)
fang_melted <- melt(fang_SNP_ID, id.vars ="SNP_ID")
fang_melted_trim <- fang_melted[,-2]

fang_group <- fang_et_al_genotypes[, -c(1:2)]
fang_t_group <- t(fang_group)
fang_t_group <- as.data.frame(fang_t_group, stringAsFactors = FALSE)
SNP_ID_group <- rownames(fang_t_group)
rownames(fang_t_group) <- NULL
fang_group <- cbind(SNP_ID_group,fang_t_group)
colnames(fang_group) <- as.character(unlist(fang_group[1,]))
fang_group = fang_group[-1, ]                                     
colnames(fang_group)[1] <- "SNP_ID"

fang_group_melt <- melt(fang_group, id.vars = "SNP_ID", measure.vars = 2:2782)
colnames(fang_group_melt)[2] <- "Group"
colnames(fang_group_melt)[3] <- "Genotype"
```

#Join the melted genotype and group data with the snp_position data
```{r}
fang_group_melt <- as.data.frame(fang_group_melt, stringAsFactors = FALSE)
joined_total <- left_join(snp_position_df, fang_group_melt, by = c("SNP_ID"))
joined_total <- joined_total[,-c(4:15)]
```

#To remove the 6 target groups
```{r}
target_groups <- c("ZMMIL", "ZMMLR","ZMMMR","ZMPBA", "ZMPIL", "ZMPJA")
joined_groups <- filter(joined_total, Group %in% target_groups)
```

#PLOTTING######################################################################
#SNPs per chromosome
# Stacked bar charts of snps per chromosome colored by group
```{r}
library(ggplot2)
qplot(factor(Chromosome), data=joined_groups, geom="bar", fill=factor(Group))
```

#Missing data and amount of heterozygosity
#Split the genotype column into two column and compare the content of the two columns
```{r}
library(tidyr)
joined_groups_split <- separate(joined_groups, Genotype, c("genotype1", "genotype2"), sep="\\/")
joined_groups_split$homozygosity <- joined_groups_split$genotype1==joined_groups_split$genotype2
```

#To encode missing genotype information as NA
```{r}
joined_groups_split[joined_groups_split=="?"]<- "NA"
```

#To alphabetically arragne by group and snp_id
```{r}
joined_groups_arrange <- joined_groups_split %>% arrange(Group,SNP_ID)
```

#To correctly encode the homozygosity column to remove false 'true' values 
```{r}
target <- c("NA")
joined_NA <- filter(joined_groups_arrange, genotype1 %in% target)
joined_NA[joined_NA=="TRUE"]<- "NA"

target_true <- c("A","T","G","C")
joined_true <- filter(joined_groups_arrange, genotype1 %in% target_true)

joined_group_new <- rbind(joined_NA, joined_true)
joined_group_new <- joined_group_new %>% arrange(Group,SNP_ID)
```

#To plot the proportion of homozygosity sites as well as missing data in each species
```{r}
ggplot(joined_group_new, aes(homozygosity, fill=Group)) + geom_bar(position = "fill") 
```

#To plot the proportion of homozygosity sites as well as missing data in each chromosome
```{r}
ggplot(joined_group_new, aes(homozygosity, fill=Chromosome)) + geom_bar(position = "fill") 
```






